lib/authorizeAccessToken.js

```diff
@@ -0,0 +1,187 @@
+ * @returns {Function} Returns an Express.js middleware route handler.
+ */
+export function authorizeAccessToken({
+  scope, cache, validIssuers, loadClientRegistration, customValidate}) {
+  return async (req, res, next) => {
+    try {
+      // Extract access token from Authorization: header (throw otherwise)
+      let clientAccessToken;
+      try {
+        // parse throws if the authorization header is missing
+        const auth = authorization.parse(req.get('authorization'));
+        if(!(auth.scheme === 'Bearer' && auth.token)) {
+          throw new Error('NotAllowedError');
+        }
+        ({token: clientAccessToken} = auth);
+      } catch(e) {
+        throw new Error('Missing or invalid "authorization" header.');
+      }
+
+      // validate and unpack the claims from the clientAccessToken
+      // use the cache as appropriate
+      // verify JWT w/cache-backed helper
+      const {valid, claims: tokenClaims, errors} = await _verifyClaimsFromJwt({
+        claimsCache: cache,
+        jwt: clientAccessToken,
+        validIssuers
+      });
+      if(!valid) {
+        const tokenError = new BedrockError(
+          'Invalid access token.', 'NotAllowedError', {
+            // errors is an array of errors which cannot be used as the
+            // BedrockError cause
+            errors,
+            httpStatusCode: 401,
+            public: true,
+          });
+        tokenError.errorId = 'invalid_token';
+        throw tokenError;
+      }
+
+      // const tokenClaims = {
+      //   iss: 'https://example.issuer.com',
+      //   client_id: 'myOauth2Client',
+      //   scope: 'read write something-else a-third-scope'
+      // };
+
+      // use the client_id claim to load the OAuth2 Client Registration
+      // Load Agent metadata based on client id
+      const {client_id} = tokenClaims;
+      let clientRegistration;
+      try {
+        (clientRegistration = await loadClientRegistration({client_id}));
+      } catch(e) {
+        throw new BedrockError(
+          'Could not load a registered agent for the provided client ID.',
+          'NotAllowedError', {
+            client_id,
+            httpStatusCode: 400,
+            public: true,
+          }, e);
+      }
+
+      // Check against provided scope
+      const {scope: tokenScope} = clientRegistration;
+      if(!(typeof tokenScope === 'string' && tokenScope.includes(scope))) {
+        throw new Error(`Access denied for token scope ${tokenScope}.` +
+          `Scope required: ${scope}.`);
+      }
+
+      // Invoke the custom authorization callback (throws error if invalid)
+      if(customValidate) {
+        await customValidate({req, claims: clientRegistration});
+      }
+
+      // if not allowed, throw a 403 Unauthorized error
+      req.clientMetadata = {...tokenClaims};
+      // otherwise, just continue
+    } catch(e) {
+      // Return the error back to the OAuth2 client
+      return next(e);
```

Comment:
this middleware seems to be mixing error handling techniques.  In some cases an error is being thrown, to be presumably be handled by a wrapping `asyncHandler`.  And here, `next(e)` is being called.  I believe this should just throw.

In order for errors to be surfaced properly to the user, they must be a `BedrockError`.